lib/vagrant/util/install_cli_autocomplete.rb

```diff
@@ -0,0 +1,119 @@
+      # This can be used to find if vagrant has inserted some strings to a file
+      #
+      # @param [string] path to a file (config file)
+      # @return [boolean] true if the prepend string is found in the file
+      def is_installed(path)
+        File.foreach(path) do |line|
+          if line.include?(@prepend_string)
+            @logger.info("Found completion already installed in #{path}")
+            return true
+          end
+        end
+        return false
+      end
+
+      # Given a path to the users home dir, will install some given strings
+      # marked by a prepend and append string
+      #
+      # @param [string] path to users home dir
+      # @return [string] path to shell config file that was modified if exists
+      def install(home)
+        path = shell_installed(home)
+        if path && !is_installed(path)
+          File.open(path, "a") do |f|
+            f.write("\n")
+            f.write(@prepend_string)
+            f.write("\n")
+            f.write(@string_insert)
+            f.write("\n")
+            f.write(@append_string)
+            f.write("\n")
+          end
+        end
+        return path
+      end
+    end
+
+    # Install autocomplete script to zsh config located as .zshrc
+    class InstallZSHShellConfig < InstallShellConfig
+      def initialize()
+        prepend_string = "# >>>> Vagrant zsh completion (start)".freeze
+        string_insert = """fpath=(#{File.join(Vagrant.source_root, "contrib", "zsh")} $fpath)\ncompinit""".freeze
+        append_string = "# <<<<  Vagrant zsh completion (end)".freeze
+        config_paths = [".zshrc".freeze].freeze
+        super(prepend_string, string_insert, append_string, config_paths)
+      end
+    end
+
+    # Install autocomplete script to bash config located as .bashrc or .bash_profile
+    class InstallBashShellConfig < InstallShellConfig
+      def initialize()
+        prepend_string = "# >>>> Vagrant bash completion (start)".freeze
+        string_insert = ". #{File.join(Vagrant.source_root, 'contrib', 'bash', 'completion.sh')}".freeze
+        append_string = "# <<<<  Vagrant bash completion (end)".freeze
+        config_paths = [".bashrc".freeze, ".bash_profile".freeze].freeze
+        super(prepend_string, string_insert, append_string, config_paths)
+      end
+    end
+
+    # Install autocomplete script for supported shells
+    class InstallCLIAutocomplete
+      SUPPORTED_SHELLS = {
+        "zsh" => Vagrant::Util::InstallZSHShellConfig.new(),
+        "bash" => Vagrant::Util::InstallBashShellConfig.new()
+      }
+
+      def self.install(shells=[])
+        shells = shells.length == 0 ? SUPPORTED_SHELLS.keys() : shells
+        home = Dir.home
+        written_paths = []
+        SUPPORTED_SHELLS.each do |k, shell|
+          if shells.include?(k)
+            p = shell.install(home)
+            if p
+              written_paths.push(p)
+            end
+          end
+        end
+        return written_paths
```

Comment:
This can be compacted down to:

```ruby
shells.map do |shell|
  next if !SUPPORTED_SHELLS[shell]
  SUPPORTED_SHELLS[shell].install(home)
end.compact
```

I'm also wondering if instead of silently ignoring requested shells that are not supported if we should be raising an `ArgumentError`.

Reply:
I think that sounds good!